src/block_normalize.rs

```diff
@@ -0,0 +1,344 @@
+
+ {
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn allocate_and_init_weights(&mut self, mi: &model_instance::ModelInstance) {
+    }
+
+    fn get_num_output_slots(&self) -> usize {1}   
+
+
+    fn get_num_output_values(&self, output: graph::OutputSlot) -> usize {
+        assert!(output.get_output_index() == 0);
+        return self.num_inputs
+    }
+    
+    fn set_input_offset(&mut self, input: graph::InputSlot, offset: usize)  {
+        assert!(input.get_input_index() == 0);
+        self.input_offset = offset;
+    }
+
+    fn set_output_offset(&mut self, output: graph::OutputSlot, offset: usize)  {
+        assert!(output.get_output_index() == 0);
+        self.output_offset = offset;
+    }
+
+
+    #[inline(always)]
+    fn forward_backward(&mut self, 
+                        further_blocks: &mut [Box<dyn BlockTrait>], 
+                        fb: &feature_buffer::FeatureBuffer, 
+                        pb: &mut port_buffer::PortBuffer, 
+                        update:bool) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+        
+        pb.tape.copy_within(self.input_offset .. (self.input_offset + self.num_inputs), self.output_offset);
+
+        block_helpers::forward_backward(further_blocks, fb, pb, update);
+                
+        if update {
+            pb.tape[self.input_offset..(self.input_offset + self.num_inputs)].fill(0.0);
+        }
+    }
+    
+    fn forward(&self, further_blocks: &[Box<dyn BlockTrait>], 
+                        fb: &feature_buffer::FeatureBuffer, 
+                        pb: &mut port_buffer::PortBuffer, 
+                        ) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+            pb.tape.copy_within(self.input_offset .. (self.input_offset + self.num_inputs), self.output_offset);
+        
+            block_helpers::forward(further_blocks, fb, pb);
+    }
+    
+}
+
+
+
+
+
+
+
+
+
+/*
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
+    use super::*;
+    use crate::block_misc;
+    use crate::feature_buffer;
+    use crate::feature_buffer::HashAndValueAndSeq;
+    use crate::vwmap;
+    use block_helpers::{slearn2, spredict2};
+    use block_misc::{Observe};
+    use crate::assert_epsilon;
+
+    fn fb_vec() -> feature_buffer::FeatureBuffer {
+        feature_buffer::FeatureBuffer {
+                    label: 0.0,
+                    example_importance: 1.0,
+                    example_number: 0,
+                    lr_buffer: Vec::new(),
+                    ffm_buffer: Vec::new(),
+                    ffm_fields_count: 0,
+        }
+    }
+
+
+    #[test]
+    fn test_simple_positive() {
+        let mut mi = model_instance::ModelInstance::new_empty().unwrap();        
```

Comment:
Is this commented out intentionally?

Reply:
it was commented out in the deep and deep-no-logs branches so i left it the same